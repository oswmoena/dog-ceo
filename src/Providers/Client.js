import axios from 'axios'

const client = axios.create({
    baseURL: `https://dog.ceo/api`,
})

export const response = async (options) => {
    options.headers = {
        'Content-Type': 'application/json',
    }

    return await client(options)
        .then(response => {
            if (response.data) {
                const { data } = response
                return data
            } else {
                return response
            }
        })
        .catch(error => {
            console.error('error', error)
        })
}