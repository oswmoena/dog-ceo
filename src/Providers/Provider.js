import { response } from './Client'

export const getAllBreeds = () => {
    const options = {
        url: `/breeds/list/all`,
    }
    return response(options)
}

export const getBreed = (breed) => {
    const options = {
        url: `/breed/${breed}/images/random/3`,
    }
    return response(options)
}

export const getSubBreed = (breed, subbreed) => {
    const options = {
        url: `/breed/${breed}/${subbreed}/images/random/3`,
    }
    return response(options)
}