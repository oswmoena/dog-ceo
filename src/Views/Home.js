import React, { useEffect, useState } from 'react'
import { getAllBreeds } from '../Providers/Provider'
import { Paper, Grid, Typography, Autocomplete, TextField } from '@mui/material'
import { createStyles, makeStyles } from '@mui/styles';
import { BreedContainer } from '../Components/BreedContainer';
import { SubBreedContainer } from '../Components/SubBreedContainer';

const useStyles = makeStyles((theme) =>
    createStyles({
        paper: {
            padding: "1% 5%",
            border: '1px solid #666666',
            color: 'black',
            fontFamily: 'Roboto-regular'
        },
        home: {
            padding: '2% 7% 0% 7%',
            backgroundColor: '#EEEEEE',
            height: '100vh'
        },
        grid: {
            paddingBottom: '10%'
        },
        title: {
            fontWeight: 'bold',
            padding: '1%',
            color: "#666666"
        },
        imageBlock: {
            maxHeight: '600px',
            overflow: 'auto'
        },
        spacing: {
            display: 'flex',
            justifyContent: 'center',
            padding: '5% 5% 0% 5%'
        },
        subSpacing: {
            display: 'flex',
            justifyContent: 'center',
            padding: '0% 5% 0% 5%'
        },
    })
);

export const Home = () => {
    const classes = useStyles();
    const [breedSelected, setBreedSelected] = useState([])
    const [subBreedSelected, setSubBreedSelected] = useState([])
    const [subBreeds, setSubBreeds] = useState([])
    const [breeds, setBreeds] = useState([])
    const [error, setError] = useState(false)

    useEffect(() => {
        bringMeBreeds();
    }, [])

    const bringMeBreeds = async () => {
        let response = await getAllBreeds()
        if (response && response.status === 'success') {
            setError(false)
            const { message } = response
            let newArrBreeds = []
            let newArrSubBreeds = []
            Object.entries(message).map(([key, val], i) => {
                if (val.length > 0) {
                    val.map(subbreed => {
                        return newArrSubBreeds = [...newArrSubBreeds, {
                            breed: key.toUpperCase(),
                            subBreed: subbreed.toUpperCase()
                        }]
                    })
                }
                return newArrBreeds = [...newArrBreeds, key.toUpperCase()]
            }
            )
            setBreeds(newArrBreeds)
            setSubBreeds(newArrSubBreeds)
        } else {
            setError(true)
        }
    }

    if (error) {
        return (
            <div >
                <div className={classes.spacing}>
                    <h3>Error en la comunicación con el servidor</h3>
                </div>
                <br />
                <div className={classes.subSpacing}>
                    <p>Por favor vuelve a intentarlo en un momento</p>
                </div>
            </div>)
    }

    return (
        <div className={classes.home}>
            <Paper elevation={2} className={classes.paper}>
                <Grid container spacing={3} justifyContent="space-between" className={classes.grid}>
                    <Grid item xs={12} md={12}>
                        <Typography variant="h3" color="initial" className={classes.title}>Dog-CEO</Typography>
                    </Grid>
                    <Grid item md={5} xs={12}>
                        <Autocomplete
                            multiple
                            id="breeds"
                            options={breeds}
                            getOptionLabel={(option) => option}
                            onChange={(event, value) => setBreedSelected(value)}
                            noOptionsText="Sin opciones para mostrar"
                            renderInput={(params) => (
                                <TextField
                                    {...params}
                                    variant="standard"
                                    label="Razas"
                                    placeholder="Elige una o más razas"
                                />
                            )}
                        />
                    </Grid>
                    <Grid item md={5} xs={12}>
                        <Autocomplete
                            multiple
                            id="sub-breeds"
                            options={subBreeds}
                            groupBy={(option) => option.breed}
                            getOptionLabel={(option) => option.subBreed}
                            onChange={(event, value) => setSubBreedSelected(value)}
                            noOptionsText="Sin opciones para mostrar"
                            renderInput={(params) => (
                                <TextField
                                    {...params}
                                    variant="standard"
                                    label="Sub Razas"
                                    placeholder="Elige una o más sub razas"
                                />
                            )}
                        />
                    </Grid>
                </Grid>
                <Grid container spacing={3} justifyContent="space-between" className={classes.imageBlock}>
                    {breedSelected.map(breed => {
                        return (
                            <Grid item md={12} xs={12} key={breed}>
                                <Typography variant="h4" color="initial" className={classes.title}>{breed}</Typography>
                                <BreedContainer breed={breed} />
                            </Grid>
                        )
                    }
                    )}
                    {subBreedSelected.map(({ breed, subBreed }) => {
                        return (
                            <Grid item md={12} xs={12} key={subBreed}>
                                <Typography variant="h4" color="initial" className={classes.title}>{breed} - {subBreed}</Typography>
                                <SubBreedContainer breed={breed} subbreed={subBreed} />
                            </Grid>
                        )
                    }
                    )}
                </Grid>


            </Paper>
        </div>
    )
}
