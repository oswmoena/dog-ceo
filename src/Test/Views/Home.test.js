import React from 'react';
import { shallow } from 'enzyme'
import { Home } from '../../Views/Home';
import { getAllBreeds } from '../../Providers/Provider';


describe('Test on Home view', () => {
    test('Show Home correctly', () => {
        const wrapper = shallow(<Home />);
        expect(wrapper).toMatchSnapshot()
    })

    test('bring all breeds', async ()=>{
        const breeds = await getAllBreeds()

        expect(breeds.length).not.toBe(0)
    })
})