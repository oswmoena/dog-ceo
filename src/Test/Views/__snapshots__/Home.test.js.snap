// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test on Home view Show Home correctly 1`] = `
<div
  className="makeStyles-home-2"
>
  <ForwardRef(Paper)
    className="makeStyles-paper-1"
    elevation={2}
  >
    <ForwardRef(Grid)
      className="makeStyles-grid-3"
      container={true}
      justifyContent="space-between"
      spacing={3}
    >
      <ForwardRef(Grid)
        item={true}
        md={12}
        xs={12}
      >
        <ForwardRef(Typography)
          className="makeStyles-title-4"
          variant="h3"
        >
          Dog-CEO
        </ForwardRef(Typography)>
      </ForwardRef(Grid)>
      <ForwardRef(Grid)
        item={true}
        md={5}
        xs={12}
      >
        <ForwardRef(Autocomplete)
          getOptionLabel={[Function]}
          id="breeds"
          multiple={true}
          noOptionsText="Sin opciones para mostrar"
          onChange={[Function]}
          options={Array []}
          renderInput={[Function]}
        />
      </ForwardRef(Grid)>
      <ForwardRef(Grid)
        item={true}
        md={5}
        xs={12}
      >
        <ForwardRef(Autocomplete)
          getOptionLabel={[Function]}
          groupBy={[Function]}
          id="sub-breeds"
          multiple={true}
          noOptionsText="Sin opciones para mostrar"
          onChange={[Function]}
          options={Array []}
          renderInput={[Function]}
        />
      </ForwardRef(Grid)>
    </ForwardRef(Grid)>
    <ForwardRef(Grid)
      className="makeStyles-imageBlock-5"
      container={true}
      justifyContent="space-between"
      spacing={3}
    />
  </ForwardRef(Paper)>
</div>
`;
